onnxruntime/contrib_ops/cuda/math/bias_softmax_impl.cu

```diff
@@ -0,0 +1,314 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/common.h"
+#include "core/providers/cuda/cuda_common.h"
+#include "core/providers/cuda/cudnn_common.h"
+#include "core/providers/cuda/cu_inc/common.cuh"
+#include "core/providers/cuda/math/softmax_impl.cuh"
+#include "core/providers/cuda/cu_inc/binary_elementwise_impl.cuh"
+#include "core/providers/cuda/math/binary_elementwise_ops_impl_functors.cuh"
+#include "bias_softmax.h"
+
+#include <limits>
+
+namespace onnxruntime {
+namespace contrib {
+namespace cuda {
+
+// Duplicated softmax_impl.cu here
+// So far attempt to use shared kernel with additional template resulted in lost performance
```

Comment:
interesting, why was the performance worse?

Reply:
It wasn't clear to me. Ke suggested the compiler index optimizations may not have taken. I would have preferred that code structure if I observed the same performance.

Reply:
Going to close this as it would require significant code change. It would better be considered when a refactor is required.